synchronize: fix for rsync protocol support

hyeonseongc opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- Bugfix Pull Request
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
2.1.0.0
```
##### SUMMARY

<!--- Describe the change, including rationale and design decisions -->

I've requested this feature in ansible/ansible#12197 and it's merged
but after I update ansible to 2.1, this feature is not working as I expected.
here's patch for fix this issue.
sorry for inconvenient

work with ansible/ansible-modules-core#4211

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

<!-- Paste verbatim command output below, e.g. before and after your change -->

```

```

hyeonseongc committed:
fix for rsync protocol support

tima commented:
~~This one looks fine to me.~~

~~shipit~~

Sorry @peter-jang had to take that back. We need some unit tests for this before it can be merged.

jctanner commented:
@peter-jang rather than forcing you to write a testcase, can you show an example task that this patch would fix?

hyeonseongc commented:
@jctanner all example tasks mentioned in https://github.com/ansible/ansible-modules-core/issues/2781 will be fixed, 
but here's some example task I've tested

pull test.

```
- name: deploy 
  hosts: myserver
  tasks:
    - name: sync files
      synchronize: >
        mode=pull delete=yes src=rsync://src.com/repo/files/ dest=/home/peter/
      delegate_to: "{{ inventory_hostname }}"

my-host$ ansible-playbook pull.yml 

PLAY [deploy] ******************************************************************

TASK [sync files] **************************************************************
changed: [10.15.20.254 -> None]

PLAY RECAP *********************************************************************
10.15.20.254               : ok=1    changed=1    unreachable=0    failed=0   
```

push test.
you must have write permission on target rsync server
and also you should check file and directory ownership on target server
because, rsync will try to sync file and directory ownership too. (with --archive option)

```
- name: deploy
  hosts: myserver
  tasks:
    - name: sync files
      synchronize: >
        mode=push delete=yes src=/home/peter/ dest=rsync://target.com/repo/files/
      delegate_to: "{{ inventory_hostname }}"

my-host$ ansible-playbook push.yml 

PLAY [deploy] ******************************************************************

TASK [sync files] **************************************************************
changed: [10.15.20.254 -> None]

PLAY RECAP *********************************************************************
10.15.20.254               : ok=1    changed=1    unreachable=0    failed=0
```

tima commented:
@peter-jang Great that you have tests however, as I understand it, they need to be unit tests. Correct jctanner?

tima commented:
Sorry for the confusion @peter-jang. I got the conversation conflated with another and misread @jctanner's comment as saying the opposite of what he said.

I'm good with this and the related ticket in core-modules. 

shipit
